src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPathExtensions.cs

```diff
@@ -101,14 +101,16 @@ public static ODataPath AppendKeySegment(this ODataPath path, IEnumerable<KeyVal
             KeySegment keySegment = new KeySegment(keys, edmType, navigationSource);
             ODataPath newPath = handler.FirstPart;
             newPath.Add(keySegment);
-            foreach (var segment in handler.LastPart)
-            {
-                newPath.Add(segment);
-            }
-
+            AppendLastSegment(handler, newPath);

```

Comment:
@marabooy By making this change, you effectively changed the logic here from `foreach (var segment in handler.LastPart)` to `foreach (var segment in handler.FirstPart)`